### ISSUE TITLE
The KERNAL should not overwrite VERA I/O

### ISSUE THREAD

**Issue by mist64** [commanderx16/x16-rom#39]
The KERNAL EDITOR assumes it owns the VERA I/O registers:

On boot, it sets up VERA_CTRL.ADDRSEL to enable VERA_DATA0, and any screen output as well as the blinking of the cursor will assume VERA_CTRL.ADDRSEL is unchanged, and it will overwrite VERA_ADDR_LO/VERA_ADDR_MID/VERA_ADDR_HI.

So changing VERA registers in BASIC direct mode has no effect, because the interrupt code will overwrite the VERA registers, and a BASIC program can't work with VERA registers and print to the screen at the same time.

Having the KERNAL save and restore the VERA I/O register state on every VERA access is too expensive. The BSOUT codepath is slow on KERNAL, even on an 8 MHz CPU, and the indirect VERA registers make it worse.

We can have the KERNAL only use VERA_DATA1 though and allow BASIC programs to own VERA_DATA0. For this, we only have to switch VERA_CTRL.ADDRSEL to 1 at the beginning (6 cycles), and restore it to 0  at the end (6 cycles) of a screen access.

The only place where we use both VERA_DATA0 and VERA_DATA1 is when scrolling. That's expensive anwyay, so we can save and resttore VERA_DATA0 here.

**Comment by jjgignac** [comment]
It's only the external registers that are of concern, right?  Why not just make a rule that the interrupt handler isn't allowed to touch VERA?  It could delegate the cursor blinking to the idle loop.

**Comment by greg-king5** [comment]
> Why not just make a rule that the interrupt handler isn't allowed to touch VERA? It could delegate the cursor blinking to the idle loop.

That wouldn't change anything.  Interrupts control the _speed_ of the blinking.  Normally, blinking is enabled during only the screen editor's idle loop.  (However, some programs want to show a blinking cursor while they manage their own typed input.  So, an interrupt-driven blink is desired.)

Mist64 was talking about the BASIC interpreter's "direct mode" (when a program isn't running).  It uses the editor for input.  The idle loop runs in between key presses.  Also, of course, printing things (including what we type) accesses VERA.

@mist64, setting `VERA_CTRL` to $00 takes only 4 cycles (a 65c02 `stz` instruction).

**Comment by jjgignac** [comment]
Let me just clarify my suggestion...

I believe flashing the cursor is the only VERA operation that currently occurs in a KERNAL interrupt handler.  I propose that the interrupt handler should continue to update address `207` ("cursor flash phase") but that a separate KERNAL subroutine (invoked from the idle loop) should update the screen state based on `207` and `206` (character at cursor position).

The downside is that programs that want to enable the cursor with `POKE204,0` would also need to `SYS` that kernel function periodically.  This strikes me as a worthwhile trade-off.

The contract would be that KERNAL subroutines that might update the screen can assume that they have complete control over VERA, and they don't need to worry about restoring the external register state when they're done.  Likewise, user programs (including BASIC programs) can trust that they have full control over VERA as long as they aren't invoking any KERNAL functions (or BASIC commands) that might update the screen.

**Comment by mist64** [comment]
I don't think this is a good tradeoff. The overhead of switching and restoring VERA_CTRL.ADDRSEL every time the cursor is inverted (twice a second or so) is minimal and does not warrant the re-architecture, and the disadvantage of not being able to turn on the cursor arbitrarily.

**Comment by codewar65** [comment]
Is this why I am unable to switch to TILEH=1 in Mode 0 on Layer 0 from BASIC to get a 8x16 bitmap font?

**Comment by mist64** [comment]
@codewar65 No, that's a known issue that is fixed in master, as well as the upcoming release.

**Comment by mist64** [comment]
General update: In current master, the interrupt handler now also moves the mouse sprite, so I added code that saves the VERA register state at the beginning of the interrupt handler, and restores it at the end.

This does not address the issue of character printing and scrolling also trashing the registers though.

**Comment by MJoergen** [comment]
What is the reason for having the VERA_CTRL.ADDRSEL register? I mean, why not have two separate I/O addresses for the two ports, i.e. 9F20-9F22 for port 0 and, say, 9F28-9F2A for port 1. From a software point of view, this would be simpler.

Making this change at this stage of the project is probably too expensive, since most of the games / demos will probably break (at least if they use address port 1). However, I don't see the reason for multiplexing the two address ports on the same I/O addresses. Could this be a limitation on the number of address lines to the VERA itself?

**Comment by mist64** [comment]
> What is the reason for having the VERA_CTRL.ADDRSEL register? I mean, why not have two separate I/O addresses for the two ports, i.e. 9F20-9F22 for port 0 and, say, 9F28-9F2A for port 1. From a software point of view, this would be simpler.

Cc @fvdhoef

**Comment by fvdhoef** [comment]
There is simply not enough address space to have the extra registers and an extra address line isn't possible since there aren't any free FPGA pins left.
